roles/openshift_health_checker/library/aos_version.py

```diff
@@ -1,91 +1,189 @@
+    def __init__(self, message, problem_pkgs=None):
+        Exception.__init__(self, message)
+        self.problem_pkgs = problem_pkgs
+
+
+def main():
+    '''Entrypoint for this Ansible module'''
     module = AnsibleModule(
         argument_spec=dict(
-            prefix=dict(required=True),  # atomic-openshift, origin, ...
-            version=dict(required=True),
+            requested_openshift_release=dict(type="str", default=''),
+            openshift_deployment_type=dict(required=True),
+            rpm_prefix=dict(required=True),  # atomic-openshift, origin, ...?
         ),
         supports_check_mode=True
     )
 
-    def bail(error):  # pylint: disable=missing-docstring
-        module.fail_json(msg=error)
+    # determine the packages we will look for
+    rpm_prefix = module.params['rpm_prefix']
+    if not rpm_prefix:
+        module.fail_json(msg="rpm_prefix must not be empty")
+    expected_pkgs = set([
+        rpm_prefix,
+        rpm_prefix + '-master',
+        rpm_prefix + '-node',
+    ])
 
-    rpm_prefix = module.params['prefix']
+    # determine what level of precision the user specified for the openshift version.
+    # should look like a version string with possibly many segments e.g. "3.4.1":
+    requested_openshift_release = str(module.params.get('requested_openshift_release', ''))
+
+    # get the list of packages available and complain if anything is wrong
+    try:
+        pkgs = _retrieve_available_packages(expected_pkgs)
+        if requested_openshift_release:
+            _check_precise_version_found(pkgs, expected_pkgs, requested_openshift_release)
+            _check_higher_version_found(pkgs, expected_pkgs, requested_openshift_release)
+        if module.params['openshift_deployment_type'] != 'origin':
+            _check_multi_minor_release(pkgs, expected_pkgs)
+    except AosVersionException as excinfo:
+        module.fail_json(msg=str(excinfo))
+    module.exit_json(changed=False)
 
-    if not rpm_prefix:
-        bail("prefix must not be empty")
 
+def _retrieve_available_packages(expected_pkgs):
+    # search for package versions available for openshift pkgs
     yb = yum.YumBase()  # pylint: disable=invalid-name
-    yb.conf.disable_excludes = ["all"]  # assume the openshift excluder will be managed, ignore current state
 
-    # search for package versions available for aos pkgs
-    expected_pkgs = [
-        rpm_prefix,
-        rpm_prefix + '-master',
-        rpm_prefix + '-node',
-    ]
+    # The openshift excluder prevents unintended updates to openshift
+    # packages by setting yum excludes on those packages. Excludes are
+    # then disabled during an install or upgrade, but this check will most
+    # likely be running outside either. When we attempt to determine what
+    # packages are available via yum they may be excluded. So, for our
+    # purposes here, disable excludes to see what will really be available
+    # during an install or upgrade.
```

Comment:
Not the way excludes are implemented currently; when enabled, they exclude the openshift packages entirely, not by version.

There's no denying the situation with excluders is a little messy, because during an install or upgrade the state changes. This should work and it's the least mess I can think of.